### PR TITLE
feat(Core/Build): add the possibility to link libraries dynamically

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- Start support `link libraries dynamically`
- https://github.com/WarheadCore/Warhead/commit/27b1930d97d541c241655d8b820cbaebe6eec44e
- https://github.com/WarheadCore/Warhead/commit/12852ddb3019aabc2f6deaff9c1b0584fd1b6bb2
- https://github.com/WarheadCore/Warhead/commit/fbc23c0713e98cb16d601c64e2e3e2a7716d2ab8

## Issues Addressed:
- Part #4302

## Target Branch(es):
- [x] Master

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/cmake/compiler/clang/settings.cmake [hunk 1]
```diff
@@ -59,6 +67,20 @@ target_compile_options(acore-compile-option-interface
...
+    target_compile_options(acore-hidden-symbols-interface
+      INTERFACE
+        -fvisibility=hidden)
+
+    # --no-undefined to throw errors when there are undefined symbols
+    # (caused through missing WARHEAD_*_API macros).
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} --no-undefined")
+
+    message(STATUS "Clang: Disallow undefined symbols")
+  endif()
```
Comment (Winfidonarleyan) [inline src/cmake/compiler/clang/settings.cmake:86]: .

File: src/cmake/compiler/msvc/settings.cmake [hunk 2]
```diff
@@ -6,9 +6,14 @@
 # set up output paths for executable binaries (.exe-files, and .dll-files on DLL-capable platforms)
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
 
-# set up output paths ofr static libraries etc (commented out - shown here as an example only)
-#set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
-#set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+set(MSVC_EXPECTED_VERSION 19.24)
+set(MSVC_EXPECTED_VERSION_STRING "Microsoft Visual Studio 2019 16.4")
```
Comment (Kitzunu) [inline src/cmake/compiler/msvc/settings.cmake:10]: good

Comment (Winfidonarleyan) [reply src/cmake/compiler/msvc/settings.cmake:10]: yes yes

Comment (Kitzunu) [reply src/cmake/compiler/msvc/settings.cmake:10]: I just said it was good O.O

Comment (Winfidonarleyan) [reply src/cmake/compiler/msvc/settings.cmake:10]: :D

**Comment by Winfidonarleyan** [comment]
Can merge

**Review by FrancescoBorzi — APPROVED** [review]

**Pull request merged**